### PR TITLE
[Bugfix] Deduplicate sampled token in top_logprobs output

### PR DESCRIPTION
**Pull request by mvanhorn**
## Purpose

Fixes #36660

When the sampled token is already among the top-k logprobs (common for greedy or near-greedy decoding), `compute_topk_logprobs()` returns the same token twice - once as the sampled token (column 0) and once in the top-k list. This causes duplicate entries in the API response's `top_logprobs` field.

## Root Cause

In `vllm/v1/worker/gpu/sample/logprob.py`, lines 103-106:

```python
logprob_token_ids = sampled_token_ids.unsqueeze(-1)  # [batch, 1]
if num_logprobs > 0:
    topk_indices = torch.topk(logits, num_logprobs, dim=-1).indices
    logprob_token_ids = torch.cat((logprob_token_ids, topk_indices), dim=1)
```

The sampled token is prepended, then `topk` tokens are appended without checking for overlap.

## Fix

- Request `num_logprobs + 1` from `torch.topk` (one extra to account for potential duplicate)
- Identify the first occurrence of the sampled token in the top-k results
- Remove it using a stable sort that preserves the original ranking order
- Slice to exactly `num_logprobs` entries

This handles both cases correctly:
- **Sampled token IS in top-k:** removed from top-k, leaving exactly `num_logprobs` unique additional entries
- **Sampled token is NOT in top-k:** no removal needed, the extra entry is simply sliced off

## Test Plan

- Verified with `ruff check` and `ruff format` (passes)
- The fix uses only standard PyTorch operations (`topk`, `gather`, `sort`) with no new dependencies
- Existing logprob tests in `tests/v1/sample/test_logprobs.py` cover the end-to-end behavior
- Manual verification: with `top_logprobs=5` and greedy sampling, the sampled token should appear only once in the response

### CODE REVIEW

File: vllm/v1/worker/gpu/sample/logprob.py [hunk 1]
```diff
@@ -102,7 +102,27 @@ def compute_topk_logprobs(
...
+        # Keep entries that are not removed.  For rows where the sampled
+        # token was NOT in the top-k, all k entries survive and we later
+        # slice to num_logprobs.  For rows where it WAS present, removing
+        # one entry leaves exactly k-1 = num_logprobs entries.
+        #
+        # Collect kept entries in order using a stable sort on the mask:
+        # kept entries get key 0, removed entries get key 1, so kept
+        # entries stay in their original relative order at the front.
+        sort_keys = remove_mask.to(torch.int32)
+        _, perm = sort_keys.sort(dim=-1, stable=True)
```
[the remark this answers is not in the document]
Comment (mvanhorn) [reply vllm/v1/worker/gpu/sample/logprob.py]: Applied the `torch.int8` suggestion in 3db7091 - saves memory since the sort keys are only 0/1 values.

**Comment by aoshen02** [comment]
Hi, problem still exists?

**Comment by mvanhorn** [comment]
Yes @aoshen02, the bug is still reproducible on current main. The PR has gone stale: vllm refactored `compute_topk_logprobs` into two paths (fast path and custom-per-request path) after this PR was opened, so the original single-site fix doesn't apply cleanly any more. I'll rebase and re-apply the dedup to both paths in a follow-up - tracking it locally. Happy to close this and open a fresh PR if that's preferred.

**Comment by mvanhorn** [comment]
Rebased onto current main and re-applied the dedup to both paths in `compute_topk_logprobs` (commit 73fd8a1). Both fast and custom paths now share a small helper (`_drop_sampled_token_from_topk`) that requests `num_logprobs + 1` from `torch.topk` and uses a stable-sort-by-mask to remove the duplicate sampled token, matching the approach reviewers accepted in the original single-site fix.

Custom path detail: the existing `_fill_logprob_token_ids_kernel` writes the (oversized) top-k columns first; the post-kernel dedup runs only on rows where `num_per_req_token_ids[req] == 0`, since rows with per-request token IDs have those columns overwritten by the kernel and don't need dedup.

Added 5 unit tests under `tests/v1/sample/test_logprobs.py`: fast-path with-dup, fast-path without-dup, custom-path with-dup, custom-path preserving per-request tokens, and `num_logprobs == 0`. Ruff clean. Tests require an accelerator so they skip on CPU; vllm CI will exercise them.

Fixes #36660. @aoshen02 thanks for the ping that surfaced the staleness.

**Comment by mvanhorn** [comment]
Addressed the DCO failure - amended the commit with Signed-off-by in 47f2bdd. The remaining two failing checks (Read the Docs build, pre-run-check) appear to be downstream of the same vllm gating workflow: pre-run-check requires a `ready` (or `verified`) label or 4+ merged author PRs in this repo (defined in .github/workflows/pre-commit.yml), and docs/pre_run_check.sh short-circuits the RTD build when pre-run-check is failing. This is my first PR to vllm, so I do not yet meet the merge-count gate. Could a maintainer apply the `ready` label so CI can run end-to-end? Happy to handle anything that surfaces after that.

**Comment by aoshen02** [comment]
I think the token-id dedup direction is right, but this seems to break rank metadata.

After this PR, if the sampled token is already top-1, `_drop_sampled_token_from_topk()` turns the candidate list from:

```text
sampled=A, topk=[A, B, C]
into:

text


[A, B, C]
However append_logprobs_for_next_position() still assigns top-k ranks as:

python

topk_ranks = range(1, num_logprobs + 1)
ranks = itertools.chain((rank,), topk_ranks)
So B will be reported as rank 1 even though it is actually rank 2, C as rank 2 even though it is actually rank 3, etc.

Could you add coverage that checks the final Logprob.rank, not only logprob_token_ids? For example, with logits [9, 8, 7, 6], sampled token 0, and num_logprobs=2, token 1 should report rank 2, not rank 1.

Also, since #36660 is an OpenAI API-visible bug, it would be good to add an e2e/API-level test that verifies top_logprobs has no duplicate sampled token in the final serialized response.

**Comment by mvanhorn** [comment]
Thanks for the precise catch. Fixed in fa0c616.

The rank fix carries each top-k token's original sorted rank through the same permutation that drops the sampled token, so survivors keep their true rank instead of the compacted index. For your example (logits [9,8,7,6], sampled 0, num_logprobs=2), token 1 now reports rank 2. The sampled token's own rank and the custom logprob_token_ids path get true ranks via a small triton kernel.

Added two unit tests asserting Logprob.rank (sampled-token-top-1 giving ranks [1,2,3], and sampled-token-outside-topk giving the sampled token rank 4 while the top-k keep [1,2]), plus an API-level test in test_chat.py asserting top_logprobs has no duplicate of the sampled token.

One caveat: the gpu/sample tests are @requires_accelerator and I don't have a GPU locally, and PR CI is still gated on the ready label, so I haven't been able to run them end to end yet. The rank math is reviewed against your worked example, but a CI pass would be the real confirmation. Happy to adjust once it can run.

**Comment by new-TonyWang** [comment]
Thanks for fixing the core issue with duplicated sampled tokens in top_logprobs. Since this PR is now quite old and does not apply cleanly to the latest main, could you please rebase it and update the implementation and tests against the current codebase? The deduplication approach itself looks good.

**Comment by mvanhorn** [comment]
Rebased onto today's main in d8d1bfeb, @new-TonyWang. The port carries exact per-token ranks (2D from the GPU worker) through the TP pack/unpack path and the rust engine-core-client protocol, which is why the diff grew relative to the old branch; empty_cpu and the legacy sampler stay 1D. CPU-runnable tests cover the dedup and ranks, and the rust protocol tests decode both shapes. The CUDA sample tests need CI.
